staging/src/k8s.io/apiserver/pkg/endpoints/filters/actas.go

```diff
@@ -0,0 +1,245 @@
+				if !groupsSpecified {
+					// if groups aren't specified for a service account, we know the groups because its a fixed mapping.  Add them
+					groups = serviceaccount.MakeGroupNames(actAsRequest.Namespace)
+				}
+
+			case v1.SchemeGroupVersion.WithKind("User").GroupKind():
+				actingAsAttributes.Resource = "users"
+				username = actAsRequest.Name
+
+			case v1.SchemeGroupVersion.WithKind("Group").GroupKind():
+				actingAsAttributes.Resource = "groups"
+				groups = append(groups, actAsRequest.Name)
+
+			case authenticationv1.SchemeGroupVersion.WithKind("UserExtra").GroupKind():
+				extraKey := actAsRequest.FieldPath
+				extraValue := actAsRequest.Name
+				actingAsAttributes.Resource = "userextras"
+				actingAsAttributes.Subresource = extraKey
+				userExtra[extraKey] = append(userExtra[extraKey], extraValue)
+
+			case authenticationv1.SchemeGroupVersion.WithKind("UID").GroupKind():
+				uid = string(actAsRequest.Name)
+				actingAsAttributes.Resource = "uids"
+
+			default:
+				klog.V(4).InfoS("unknown actas request type", "request", actAsRequest)
+				responsewriters.Forbidden(ctx, actingAsAttributes, w, req, fmt.Sprintf("unknown actas request type: %v", actAsRequest), s)
+				return
+			}
+
+			decision, reason, err := a.Authorize(ctx, actingAsAttributes)
+			if err != nil || decision != authorizer.DecisionAllow {
+				klog.V(4).InfoS("Forbidden", "URI", req.RequestURI, "reason", reason, "err", err)
+				responsewriters.Forbidden(ctx, actingAsAttributes, w, req, reason, s)
+				return
+			}
+		}
+
+		authorized, reason, err := a.Authorize(ctx, attributes)
+		if err != nil || authorized != authorizer.DecisionAllow {
+			klog.V(4).InfoS("Forbidden", "URI", req.RequestURI, "reason", reason, "err", err)
+			responsewriters.Forbidden(ctx, attributes, w, req, reason, s)
+			return
+		}
+
+		if username != user.Anonymous {
+			// When acting as a non-anonymous user, include the 'system:authenticated' group
+			// in the acted user info:
+			// - if no groups were specified
+			// - if a group has been specified other than 'system:authenticated'
+			//
+			// If 'system:unauthenticated' group has been specified we should not include
+			// the 'system:authenticated' group.
+			addAuthenticated := true
+			for _, group := range groups {
+				if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+					addAuthenticated = false
+					break
+				}
+			}
+
+			if addAuthenticated {
+				groups = append(groups, user.AllAuthenticated)
+			}
+		} else {
+			addUnauthenticated := true
+			for _, group := range groups {
+				if group == user.AllUnauthenticated {
+					addUnauthenticated = false
+					break
+				}
+			}
+
+			if addUnauthenticated {
+				groups = append(groups, user.AllUnauthenticated)
+			}
+		}
+
+		newUser := &user.DefaultInfo{
+			Name:   username,
+			Groups: groups,
+			Extra:  userExtra,
+			UID:    uid,
+		}
+		req = req.WithContext(request.WithUser(ctx, newUser))
```

Comment:
replacing the user with the act-as user isn't intended.  We want the user to remain the user who made the request.